### PR TITLE
chore(asm): response headers collection

### PR DESCRIPTION
**Pull request by vdeturckheim**
This PR does two things:
* introduce response headers collection in case of attacks
* ensures we collect request headers in case of attacks even if they are not requested by the rules

### CODE REVIEW

File: ddtrace/appsec/processor.py [hunk 1]
```diff
@@ -86,15 +86,21 @@ class _Addresses(object):
...
+    "content-length",
+    "content-type",
+    "Content-Encoding",
+    "Content-Language",
+}
 
 
-def _set_headers(span, headers):
-    # type: (Span, Dict[str, Union[str, List[str]]]) -> None
+def _set_headers(span, kind, to_collect, headers):
```
Comment (P403n1x87) [inline ddtrace/appsec/processor.py:98]: If we expect `to_collect` to essentially default to `_COLLECTED_RESPONSE_HEADERS`, I would move the global into a local for a slightly faster look up, and also make it a `frozenset`. WDYT?

Comment (vdeturckheim) [reply ddtrace/appsec/processor.py:98]: indeed, `_COLLECTED_RESPONSE_HEADERS` is a subset of `_COLLECTED_REQUEST_HEADERS`. But We call the method with both for each attack, would there still be a gain?
+1 for `frozenset`

File: ddtrace/appsec/processor.py [hunk 2]
```diff
@@ -90,18 +90,19 @@ class _Addresses(object):
...
-def _set_headers(span, kind, to_collect, headers):
-    # type: (Span, str, Set[str], Dict[str, Union[str, List[str]]]) -> None
+def _set_headers(span, kind, headers, extra_collect={}):
+    # type: (Span, str, Dict[str, Union[str, List[str]]], Set[str]) -> None
+    always_collect = frozenset(
+        {
+            "content-length",
+            "content-type",
+            "Content-Encoding",
+            "Content-Language",
```
Comment (vdeturckheim) [inline ddtrace/appsec/processor.py:100]:
```suggestion
            "content-encoding",
            "content-language",
```

[the remark this answers is not in the document]
Comment (vdeturckheim) [reply ddtrace/appsec/processor.py:100]: Good catch 😅

File: ddtrace/appsec/processor.py [hunk 3]
```diff
@@ -89,15 +89,22 @@ class _Addresses(object):
...
 }
 
-_COLLECTED_HEADER_PREFIX = "http.request.headers."
 
-
-def _set_headers(span, headers):
-    # type: (Span, Dict[str, Union[str, List[str]]]) -> None
+def _set_headers(span, kind, headers, extra_collect={}):
+    # type: (Span, str, Dict[str, Union[str, List[str]]], Set[str]) -> None
+    always_collect = frozenset(
```
Comment (jd) [inline ddtrace/appsec/processor.py:95]: maybe put that outside the function as a global/constant?

File: ddtrace/appsec/processor.py [hunk 4]
```diff
@@ -89,15 +89,22 @@ class _Addresses(object):
     "x-real-ip",
 }
 
-_COLLECTED_HEADER_PREFIX = "http.request.headers."
 
-
-def _set_headers(span, headers):
-    # type: (Span, Dict[str, Union[str, List[str]]]) -> None
+def _set_headers(span, kind, headers, extra_collect={}):
```
Comment (jd) [inline ddtrace/appsec/processor.py:93]:
```suggestion
def _set_headers(span, kind, headers, extra_collect=None):
```
Never use `{}` as a default value for a kwarg, it's dangerous. It's going to be the same object for every call.
It's an antipattern in Python.

**Review by jd — CHANGES_REQUESTED** [review]

File: ddtrace/appsec/processor.py (hunk 4 above)
Comment (vdeturckheim) [reply ddtrace/appsec/processor.py:93]: Gotcha! Thanks

File: ddtrace/appsec/processor.py [hunk 5]
```diff
@@ -68,36 +69,47 @@ class _Addresses(object):
...
+    }
+)
+
+
+_ALWAYS_COLLECT_HEADERS = frozenset(
+    {
+        "content-length",
+        "content-type",
+        "Content-Encoding",
+        "Content-Language",
```
Comment (Moutix) [inline ddtrace/appsec/processor.py:101]:
```suggestion
        "content-encoding",
        "content-language",
```

File: ddtrace/appsec/processor.py [hunk 6]
```diff
@@ -68,36 +69,47 @@ class _Addresses(object):
...
+    }
+)
+
+
+def _set_headers(span, kind, headers, extra_collect=None):
+    # type: (Span, str, Dict[str, Union[str, List[str]]], Optional[frozenset[str]]) -> None
     for k in headers:
-        if k.lower() in _COLLECTED_REQUEST_HEADERS:
+        low = k.lower()
+        if low in _ALWAYS_COLLECT_HEADERS or extra_collect is not None and low in extra_collect:
```
Comment (Moutix) [inline ddtrace/appsec/processor.py:110]:
```suggestion
        if low in _ALWAYS_COLLECT_HEADERS or low in (extra_collect or {}):
```

File: ddtrace/appsec/processor.py (hunk 5 above)
Comment (vdeturckheim) [reply ddtrace/appsec/processor.py:101]: 🤦  on me

**Comment by vdeturckheim** [comment]
Gunicorn seems to make the apps parsing the headers differently. I will move this in draft to take the time to understand this

File: ddtrace/appsec/processor.py (hunk 3 above)
Comment (vdeturckheim) [reply ddtrace/appsec/processor.py]: Done

File: ddtrace/appsec/processor.py (hunk 4 above)
Comment (vdeturckheim) [reply ddtrace/appsec/processor.py]: Fixed

**Review by mabdinur — APPROVED** [review]

**Comment by vdeturckheim** [comment]
@jd, should be good :)

**Review by P403n1x87 — DISMISSED** [review]

File: ddtrace/appsec/processor.py [hunk 7]
```diff
@@ -183,10 +195,11 @@ def on_span_finish(self, span):
             if request_query is not None:
                 data[_Addresses.SERVER_REQUEST_QUERY] = request_query
 
-        if self._is_needed(_Addresses.SERVER_REQUEST_HEADERS_NO_COOKIES):
-            request_headers = _context.get_item("http.request.headers", span=span)
-            if request_headers is not None:
-                data[_Addresses.SERVER_REQUEST_HEADERS_NO_COOKIES] = _transform_headers(request_headers)
+        request_headers = _context.get_item("http.request.headers", span=span)
+        if request_headers is not None:
+            request_headers = _transform_headers(request_headers)
```
[the remark this answers is not in the document]
Comment (vdeturckheim) [reply ddtrace/appsec/processor.py:200]: We'll need it for reporting later too, that's why I moved it out

File: ddtrace/appsec/processor.py [hunk 8]
```diff
@@ -68,36 +69,47 @@ class _Addresses(object):
...
+    }
+)
+
+
+def _set_headers(span, kind, headers, extra_collect=None):
+    # type: (Span, str, Dict[str, Union[str, List[str]]], Optional[frozenset[str]]) -> None
     for k in headers:
-        if k.lower() in _COLLECTED_REQUEST_HEADERS:
+        low = k.lower()
+        if low in _ALWAYS_COLLECT_HEADERS or low in (extra_collect or {}):
```
Comment (P403n1x87) [inline ddtrace/appsec/processor.py:110]: can we do `extra_collect or {}` outside the loop, e.g.
~~~ python
extra_collect = extra_collect or {}
~~~
just before the `for`?

**Pull request closed without merging**